File: 314-binary-tree-vertical-order-traversal/314-binary-tree-vertical-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        col_map, q = defaultdict(list), [(root, 0)]
        min_col, max_col = float('inf'), float('-inf')
        while q:
            nq = []
            for node, col in q:
                col_map[col].append(node.val)
                min_col = min(min_col, col)
                max_col = max(max_col, col)
                
                if node.left:
                    nq.append((node.left, col - 1))
                if node.right:
                    nq.append((node.right, col + 1))
            q = nq
            
        return [col_map[col] for col in range(min_col, max_col + 1)]
        
        

File: 314-binary-tree-vertical-order-traversal/314-binary-tree-vertical-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class NodeColumn {
    public TreeNode node;
    public int col;
    
    NodeColumn(TreeNode node, int col){
        this.node = node;
        this.col = col;
    }
}

class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        if(root == null){
            return new ArrayList<List<Integer>>();
        }
        
        Map<Integer, List<Integer>> colValues = new HashMap<>();
        
        List<NodeColumn> q = new ArrayList<>();
        q.add(new NodeColumn(root, 0));
        
        int minCol = Integer.MAX_VALUE, maxCol = Integer.MIN_VALUE;
        while(q.size() > 0){
            List<NodeColumn> nq = new ArrayList<>();
            for(NodeColumn nc : q){
                if(!colValues.containsKey(nc.col)){
                    colValues.put(nc.col, new ArrayList<>());
                }
                colValues.get(nc.col).add(nc.node.val);
                minCol = Math.min(minCol, nc.col);
                maxCol = Math.max(maxCol, nc.col);
                
                if(nc.node.left != null){
                    nq.add(new NodeColumn(nc.node.left, nc.col - 1));
                }
                
                if(nc.node.right != null){
                    nq.add(new NodeColumn(nc.node.right, nc.col + 1));
                }
            }
            q = nq;
        }
        
        List<List<Integer>> ans = new ArrayList<>();
        for(int i = minCol; i <= maxCol; ++i){
            ans.add(colValues.get(i));
        }
        return ans;
    }
}